utils/src/lib.rs

```diff
@@ -531,123 +531,101 @@ pub fn encode_args(func: &Function, args: &[impl AsRef<str>]) -> Result<Vec<u8>>
     Ok(func.encode_input(&tokens)?)
 }
 
-/// Fetches a function signature given the selector using 4byte.directory
-pub async fn fourbyte(selector: &str) -> Result<Vec<(String, i32)>> {
+pub enum SelectorType {
+    Function,
+    Event,
+}
+
+/// Decodes the given function or event selector using sig.eth.samczsun.com
+async fn decode_selector(selector: &str, selector_type: SelectorType) -> Result<Vec<String>> {
     #[derive(Deserialize)]
     struct Decoded {
-        text_signature: String,
-        id: i32,
+        name: String,
     }
 
     #[derive(Deserialize)]
-    struct ApiResponse {
-        results: Vec<Decoded>,
+    struct ApiResult {
+        event: HashMap<String, Vec<Decoded>>,
+        function: HashMap<String, Vec<Decoded>>,
     }
 
-    let selector = &selector.strip_prefix("0x").unwrap_or(selector);
-    if selector.len() < 8 {
-        return Err(eyre::eyre!("Invalid selector"))
+    #[derive(Deserialize)]
+    struct ApiResponse {
+        ok: bool,
+        result: ApiResult,
     }
-    let selector = &selector[..8];
 
-    let url = format!("https://www.4byte.directory/api/v1/signatures/?hex_signature={selector}");
-    let res = reqwest::get(url).await?;
-    let res = res.text().await?;
+    // using samczsun signature database over 4byte
+    // see https://github.com/foundry-rs/foundry/issues/1672
+    let base_url = "https://sig.eth.samczsun.com/api/v1/signatures";
+    let url = match selector_type {
+        SelectorType::Function => format!("{base_url}?function={selector}"),
+        SelectorType::Event => format!("{base_url}?event={selector}"),
+    };
+
+    let res = reqwest::get(url).await?.text().await?;
     let api_response = match serde_json::from_str::<ApiResponse>(&res) {
         Ok(inner) => inner,
         Err(err) => {
             eyre::bail!("Could not decode response:\n {res}.\nError: {err}")
         }
     };
 
-    if api_response.results.is_empty() {
-        eyre::bail!("no signature found for provided function selector")
+    if !api_response.ok {
+        eyre::bail!("Failed to decode:\n {res}")
     }
 
-    Ok(api_response
-        .results
+    let decoded = match selector_type {
+        SelectorType::Function => api_response.result.function,
+        SelectorType::Event => api_response.result.event,
+    };
+
+    Ok(decoded
+        .get(selector)
+        .ok_or(eyre::eyre!("No signature found"))?
         .into_iter()
-        .map(|d| (d.text_signature, d.id))
-        .collect::<Vec<(String, i32)>>())
+        .map(|d| d.name.clone())
+        .collect::<Vec<String>>())
```

Comment:
a `remove` would equivalent here?

Reply:
I apologize - not sure I follow. `remove` instead of `get` from the map to avoid the clone?